File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Public/Extensions/Hybrid/HybridSTTRouter.ts
/**
 * HybridSTTRouter.ts
 *
 * THIN React Native binding over the commons STT hybrid router
 * (rac_stt_hybrid_router + its proto-byte ABI), reached through the Nitro
 * `RunAnywhereCore` hybrid object's `hybridSttRouter*` / `hybrid*` methods.
 * Per-request dispatch between an on-device (offline, sherpa) STT service and a
 * cloud (online, cloud) STT service.
 *
 * Division of labour — commons owns ALL routing: the filter phase, rank/sort,
 * confidence cascade, and primary->secondary fallback all live in
 * rac_stt_hybrid_router.cpp. NONE of that logic is reimplemented here. This
 * binding only:
 *   1. creates the router handle,
 *   2. creates the two STT services through the registry-routed creation path
 *      (hint "sherpa"/"cloud"; provider in the cloud config) and attaches
 *      them with their `HybridModelDescriptor` bytes,
 *   3. registers any custom-filter predicates (JS->native) and installs the
 *      `HybridRoutingPolicy` policy bytes,
 *   4. drives transcribe and decodes the `HybridSttTranscribeResponse`.
 *
 * Mirrors Swift `HybridSTTRouter` + Kotlin `RACRouter.SttRouter`. Lifetime: the
 * router does NOT own the underlying services. This class tracks each service
 * handle, clears the router slots before destroying the services (the
 * use-after-free guard called out in rac_stt_hybrid_router.h), unregisters its
 * custom filters, and tears everything down in `close()`.
 */

import { requireNativeModule, isNativeModuleAvailable } from '../../../native';
import { ensureServicesReady } from '../../../Foundation/Initialization/ServicesReadyGuard';
import { SDKLogger } from '../../../Foundation/Logging/Logger/SDKLogger';
import { SDKException } from '../../../Foundation/Errors/SDKException';
import { ErrorCategory, ErrorCode } from '@runanywhere/proto-ts/errors';
import { arrayBufferToBytes, bytesToArrayBuffer } from '../../../services/ProtoBytes';
import { encodeProtoMessage } from '../../../services/ProtoWire';
import {
  HybridModelDescriptor,
  HybridSttTranscribeRequest,
  HybridSttTranscribeResponse,
} from '@runanywhere/proto-ts/hybrid_router';
import {
  HybridBackendKind,
  pinnedEngineName,
  type HybridModel,
  type HybridTranscribeOptions,
  type HybridTranscribeResult,
} from './HybridModel';
import {
  encodeHybridRoutingPolicy,
  customFiltersOf,
  type HybridRoutingPolicy,
} from './HybridRoutingPolicy';
import { CloudSTT } from './CloudSTT';

const logger = new SDKLogger('Hybrid.STTRouter');

const RAC_SUCCESS = 0;

/** One attached service: the native handle plus the descriptor used to attach it. */
interface AttachedService {
  handle: number;
  model: HybridModel;
}

/**
 * A hybrid STT router pairing one offline + one online speech service.
 *
 * Usage:
 * ```typescript
 * await CloudSTT.register();
 * await CloudSTT.registerModel({ id: 'saaras', provider: 'sarvam',
 *   model: 'saaras:v2.5', apiKey: '…' });
 *
 * const router = await HybridSTTRouter.create();
 * await router.setPair(
 *   offlineSherpa('sherpa-onnx-whisper-tiny.en'),
 *   onlineCloud('saaras'),
 *   { hardFilters: [Filters.network()], cascade: Cascades.confidence(0.5),
 *     rank: HybridRank.HYBRID_RANK_PREFER_LOCAL_FIRST }
 * );
 * const result = await router.transcribe(audio, { audioFormat: 1 });
 * await router.close();
 * ```
 */
export class HybridSTTRouter {
  private handle: number;
  private offline: AttachedService | null = null;
  private online: AttachedService | null = null;
  private customFilterNames: string[] = [];

  private constructor(handle: number) {
    this.handle = handle;
  }

  /** Create the native router handle (`rac_stt_hybrid_router_create`). */
  static async create(): Promise<HybridSTTRouter> {
    if (!isNativeModuleAvailable()) {
      throw SDKException.nativeModuleUnavailable();
    }
    await ensureServicesReady();
    const handle = await requireNativeModule().hybridSttRouterCreate();
    if (!handle) {
      // Swift parity: create failure throws .serviceNotAvailable
      // (HybridSTTRouter.swift:97-103).
      throw SDKException.serviceNotAvailable('rac_stt_hybrid_router_create returned 0');
    }
    return new HybridSTTRouter(handle);
  }

  /**
   * Bind the offline + online models, install the policy, and register any
   * custom-filter predicates. Replaces any previous pairing.
   */
  async setPair(
    offline: HybridModel,
    online: HybridModel,
    policy: HybridRoutingPolicy = {}
  ): Promise<void> {
    this.ensureOpen();
    const native = requireNativeModule();

    // Build both services up-front so a failure on the online side doesn't
    // leave a half-attached router.
    const offlineHandle = await this.createService(offline);
    let onlineHandle: number;
    try {
      onlineHandle = await this.createService(online);
    } catch (error) {
      await native.hybridSttRouterDestroyService(offlineHandle);
      throw error;
    }

    // Detach + destroy any previously attached services before swapping in the
    // new pair (clear router slots first — header UAF note), and retire the
    // previous policy's custom-filter predicates.
    await this.clearAndDestroyServices();
    await this.unregisterCustomFilters();

    // Attach both sides with their descriptor bytes.
    const rcOff = await native.hybridSttRouterSetOfflineService(
      this.handle,
      offlineHandle,
      this.descriptorBytes(offline)
    );
    if (rcOff !== RAC_SUCCESS) {
      await native.hybridSttRouterDestroyService(offlineHandle);
      await native.hybridSttRouterDestroyService(onlineHandle);
      // Swift parity: HybridSTTRouter.swift:166-173 throws .serviceNotAvailable.
      throw SDKException.serviceNotAvailable(
        `hybridSttRouterSetOfflineService rc=${rcOff}`
      );
    }
    const rcOn = await native.hybridSttRouterSetOnlineService(
      this.handle,
      onlineHandle,
      this.descriptorBytes(online)
    );
    if (rcOn !== RAC_SUCCESS) {
      await native.hybridSttRouterSetOfflineService(this.handle, 0, this.emptyBytes());
      await native.hybridSttRouterDestroyService(offlineHandle);
      await native.hybridSttRouterDestroyService(onlineHandle);
      // Swift parity: HybridSTTRouter.swift:178-186 throws .serviceNotAvailable.
      throw SDKException.serviceNotAvailable(
        `hybridSttRouterSetOnlineService rc=${rcOn}`
      );
    }

    // Register each custom filter's predicate by NAME with the cross-SDK commons
    // callback table BEFORE installing the policy bytes, so the router can
    // resolve each name the first time it filters. Commons evaluates them —
    // this layer does NOT pre-filter candidates or toggle slots.
    const customs = customFiltersOf(policy);
    for (const custom of customs) {
      const rc = await native.hybridRegisterCustomFilter(
        custom.name,
        async (candidateModelId: string) =>
          Promise.resolve(custom.check(candidateModelId))
      );
      if (rc !== RAC_SUCCESS) {
        // Roll back: detach services + unregister the filters we just added.
        for (const c of customs) {
          if (c.name === custom.name) break;
          await native.hybridUnregisterCustomFilter(c.name);
        }
        await native.hybridSttRouterSetOfflineService(this.handle, 0, this.emptyBytes());
        await native.hybridSttRouterSetOnlineService(this.handle, 0, this.emptyBytes());
        await native.hybridSttRouterDestroyService(offlineHandle);
        await native.hybridSttRouterDestroyService(onlineHandle);
        // Swift parity: registration failures during setPair throw
        // .serviceNotAvailable (HybridSTTRouter.swift:200-210).
        throw SDKException.serviceNotAvailable(
          `hybridRegisterCustomFilter('${custom.name}') rc=${rc}`
        );
      }
    }

    const rcPolicy = await native.hybridSttRouterSetPolicy(
      this.handle,
      encodeHybridRoutingPolicy(policy)
    );
    if (rcPolicy !== RAC_SUCCESS) {
      for (const c of customs) await native.hybridUnregisterCustomFilter(c.name);
      await native.hybridSttRouterSetOfflineService(this.handle, 0, this.emptyBytes());
      await native.hybridSttRouterSetOnlineService(this.handle, 0, this.emptyBytes());
      await native.hybridSttRouterDestroyService(offlineHandle);
      await native.hybridSttRouterDestroyService(onlineHandle);
      // Swift parity: HybridSTTRouter.swift:200-210 throws .serviceNotAvailable.
      throw SDKException.serviceNotAvailable(`hybridSttRouterSetPolicy rc=${rcPolicy}`);
    }

    this.offline = { handle: offlineHandle, model: offline };
    this.online = { handle: onlineHandle, model: online };
    this.customFilterNames = customs.map((c) => c.name);
  }

  /**
   * Run one transcribe request through the router. The router applies the
   * installed policy (filters -> rank -> invoke -> fallback) in commons and
   * returns the chosen backend's result plus the routing decision.
   *
   * @param audio   File-encoded audio (wav/mp3/flac/…) OR raw PCM bytes. Raw
   *                PCM16 is wrapped in a WAV container by the commons router
   *                (rac_stt_hybrid_router_proto.cpp) so one payload serves
   *                both services; WAV input and declared compressed formats
   *                pass through unchanged.
   * @param options Optional language / sampleRate / audioFormat hints.
   */
  async transcribe(
    audio: Uint8Array,
    options: HybridTranscribeOptions = {}
  ): Promise<HybridTranscribeResult> {
    this.ensureOpen();
    if (this.offline == null || this.online == null) {
      // No direct Swift analog (Swift lets commons reject the unpaired
      // transcribe); aligned with the Swift router's .serviceNotAvailable
      // failure family for missing services.
      throw SDKException.serviceNotAvailable('setPair() not called');
    }
    const native = requireNativeModule();

    const request = HybridSttTranscribeRequest.fromPartial({
      audioBytes: audio,
      context: {},
      options: {
        language: options.language ?? '',
        sampleRate: options.sampleRate ?? 0,
        audioFormat: options.audioFormat ?? 0,
      },
    });
    const responseBytes = await native.hybridSttRouterTranscribe(
      this.handle,
      encodeProtoMessage(request, HybridSttTranscribeRequest)
    );
    return this.decodeResponse(responseBytes);
  }

  /**
   * Best-effort cancel of an in-flight transcribe. No STT engine exposes a
   * cancel op today, so commons treats this as a no-op until one does.
   */
  async cancel(): Promise<void> {
    if (!this.handle) return;
    await requireNativeModule().hybridSttRouterCancel(this.handle);
  }

  /**
   * Detach + destroy both services, unregister custom filters, and destroy the
   * router handle. Idempotent.
   */
  async close(): Promise<void> {
    if (!this.handle) return;
    const native = requireNativeModule();
    await this.unregisterCustomFilters();
    await this.clearAndDestroyServices();
    await native.hybridSttRouterDestroy(this.handle);
    this.handle = 0;
  }

  // --- internals ------------------------------------------------------------

  private ensureOpen(): void {
    if (!this.handle) {
      // Swift parity: notOpen() throws .notInitialized with category
      // .component and message "HybridSTTRouter is closed"
      // (HybridSTTRouter.swift:490-496).
      throw SDKException.of(
        ErrorCode.ERROR_CODE_NOT_INITIALIZED,
        'HybridSTTRouter is closed',
        { category: ErrorCategory.ERROR_CATEGORY_COMPONENT }
      );
    }
  }

  /**
   * Create one registry-routed STT service for `model`. cloud needs the
   * provider + api_key + model from the credential registry (resolved here via
   * CloudSTT.configJSON); sherpa resolves its model from the C model registry,
   * so it gets the model id with no extra config.
   */
  private async createService(model: HybridModel): Promise<number> {
    const native = requireNativeModule();
    const engineHint = pinnedEngineName(model.backend);
    const isCloud = model.backend === HybridBackendKind.HYBRID_BACKEND_CLOUD;
    const configJson = isCloud ? CloudSTT.configJSON(model.id) : '';
    const modelIdOrPath = isCloud ? '' : model.id;
    const handle = await native.hybridSttRouterCreateService(
      engineHint,
      modelIdOrPath,
      configJson
    );
    if (!handle) {
      // Swift parity: createService failures throw .serviceNotAvailable
      // (HybridSTTRouter.swift:368-409).
      throw SDKException.serviceNotAvailable(
        `Failed to create '${engineHint}' STT service for model '${model.id}'. ` +
          'Register the backend first (ONNX.register() for sherpa, ' +
          'CloudSTT.register() for cloud).'
      );
    }
    return handle;
  }

  /** Encode a `HybridModelDescriptor` for the router setters. */
  private descriptorBytes(model: HybridModel): ArrayBuffer {
    const descriptor = HybridModelDescriptor.fromPartial({
      modelId: model.id,
      modelType: model.modelType,
      backend: model.backend,
      provider: model.provider,
    });
    return encodeProtoMessage(descriptor, HybridModelDescriptor);
  }

  private emptyBytes(): ArrayBuffer {
    return bytesToArrayBuffer(new Uint8Array(0));
  }

  /**
   * Clear both router slots, then destroy whatever services were attached.
   * Slot-clearing must precede service destruction (the router holds raw
   * pointers — see rac_stt_hybrid_router.h UAF note).
   */
  private async clearAndDestroyServices(): Promise<void> {
    const native = requireNativeModule();
    if (this.handle) {
      await native.hybridSttRouterSetOfflineService(this.handle, 0, this.emptyBytes());
      await native.hybridSttRouterSetOnlineService(this.handle, 0, this.emptyBytes());
    }
    const offline = this.offline;
    const online = this.online;
    this.offline = null;
    this.online = null;
    if (offline) await native.hybridSttRouterDestroyService(offline.handle);
    if (online) await native.hybridSttRouterDestroyService(online.handle);
  }

  /** Unregister every custom-filter predicate this router installed. Idempotent. */
  private async unregisterCustomFilters(): Promise<void> {
    const native = requireNativeModule();
    const names = this.customFilterNames;
    this.customFilterNames = [];
    for (const name of names) {
      await native.hybridUnregisterCustomFilter(name);
    }
  }

  /** Decode a `HybridSttTranscribeResponse`, raising the native rc as an exception. */
  private decodeResponse(buffer: ArrayBuffer): HybridTranscribeResult {
    const bytes = arrayBufferToBytes(buffer);
    if (bytes.byteLength === 0) {
      // Swift parity: transcribe failures throw .serviceNotAvailable
      // (HybridSTTRouter.swift:256-262).
      throw SDKException.serviceNotAvailable('Hybrid STT transcribe returned no response');
    }
    const message = HybridSttTranscribeResponse.decode(bytes);
    if (message.rc !== RAC_SUCCESS) {
      const reason = message.errorMsg || `Hybrid STT transcribe failed (rc=${message.rc})`;
      logger.warning(reason);
      // Swift parity: decodeResponse throws .serviceNotAvailable
      // (HybridSTTRouter.swift:291-300).
      throw SDKException.serviceNotAvailable(reason);
    }
    const routing = message.routing;
    return {
      text: message.text,
      detectedLanguage: message.detectedLanguage,
      routing: {
        chosenModelId: routing?.chosenModelId ?? '',
        wasFallback: routing?.wasFallback ?? false,
        attemptCount: routing?.attemptCount ?? 0,
        primaryErrorCode: routing?.primaryErrorCode ?? 0,
        primaryErrorMessage: routing?.primaryErrorMessage ?? '',
        confidence: routing?.confidence ?? Number.NaN,
        primaryConfidence: routing?.primaryConfidence ?? Number.NaN,
      },
    };
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/public/hybrid/hybrid_stt_router.dart
// SPDX-License-Identifier: Apache-2.0
//
// hybrid_stt_router.dart — THIN Dart binding over the commons STT hybrid router
// (rac_stt_hybrid_router + its proto-byte ABI). Per-request dispatch between an
// on-device (offline, sherpa) STT service and a cloud (online, cloud) STT
// service.
//
// Division of labour — commons owns ALL routing: filter phase, rank/sort,
// confidence cascade, and primary→secondary fallback all live in
// rac_stt_hybrid_router.cpp. NONE of that logic is reimplemented here. This
// binding only:
//   1. creates the router handle,
//   2. creates the two STT services through the registry-routed creation path
//      (rac_plugin_find_for_engine(TRANSCRIBE, "sherpa"/"cloud") →
//      stt_ops->create) and attaches them with their HybridModelDescriptor bytes,
//   3. registers any custom-filter predicates and installs the policy bytes,
//   4. drives the router's transcribe and decodes the response.
//
// Mirrors the Kotlin RACRouter.SttRouter + HybridRouterBridgeAdapter and the
// Swift HybridSTTRouter feature surface. All FFI lives in
// `lib/native/dart_bridge_hybrid_stt.dart` (canonical §15 type-discipline).
//
// Lifetime: the router does NOT own the underlying services. This object keeps
// each service in stable native storage for the router's lifetime, clears the
// router slots BEFORE destroying the services (avoiding the use-after-free
// called out in rac_stt_hybrid_router.h), and tears everything down in [close].

import 'dart:typed_data';

import 'package:runanywhere/generated/hybrid_router.pb.dart' as pb;
import 'package:runanywhere/native/dart_bridge_hybrid_stt.dart';
import 'package:runanywhere/native/types/basic_types.dart'
    show RacHandle, RAC_SUCCESS;
import 'package:runanywhere/public/hybrid/hybrid_cloud_backend.dart';
import 'package:runanywhere/public/hybrid/hybrid_model.dart';
import 'package:runanywhere/public/hybrid/hybrid_routing_policy.dart';

/// A hybrid STT router pairing one offline + one online speech service.
///
/// Usage:
/// ```dart
/// RunAnywhere.hybrid.registerCloud();                       // fold cloud plugin in
/// RunAnywhere.hybrid.cloud.register(                         // provider as data
///   id: 'saaras', provider: 'sarvam', model: 'saaras:v2.5', apiKey: '…');
/// RunAnywhere.hybrid.setDeviceStateProvider(myProvider);     // optional NETWORK/Battery
///
/// final router = RunAnywhere.hybrid.createSttRouter();
/// router.setPair(
///   offline: HybridModel.offlineSherpa('sherpa-onnx-whisper-tiny.en'),
///   online:  HybridModel.onlineCloud('saaras'),
///   policy:  const HybridRoutingPolicy(
///     hardFilters: [HybridFilter.network()],
///     cascade: HybridCascade.confidence(0.5),
///   ),
/// );
/// final result = router.transcribe(audioBytes,
///     options: pb.HybridSttTranscribeOptions(audioFormat: 1));
/// router.close();
/// ```
class HybridSttRouter {
  HybridSttRouter._(this._handle);

  final DartBridgeHybridStt _bridge = DartBridgeHybridStt.instance;

  RacHandle _handle;
  HybridServiceHandle? _offline;
  HybridServiceHandle? _online;
  List<String> _customFilterNames = const <String>[];
  bool _closed = false;

  /// Allocate a native router handle. Throws [StateError] on failure.
  /// Prefer `RunAnywhere.hybrid.createSttRouter()`.
  factory HybridSttRouter.create() =>
      HybridSttRouter._(DartBridgeHybridStt.instance.createRouter());

  /// Bind the offline + online models, install the [policy], and register any
  /// custom-filter predicates. Replaces any previous pairing.
  ///
  /// [offline] must be an [HybridModelKind.offline] model and [online] an
  /// [HybridModelKind.online] model; otherwise an [ArgumentError] is thrown.
  void setPair({
    required HybridModel offline,
    required HybridModel online,
    HybridRoutingPolicy policy = const HybridRoutingPolicy(),
  }) {
    _ensureOpen();
    if (offline.kind != HybridModelKind.offline) {
      throw ArgumentError.value(
          offline.kind, 'offline', 'must be HybridModelKind.offline');
    }
    if (online.kind != HybridModelKind.online) {
      throw ArgumentError.value(
          online.kind, 'online', 'must be HybridModelKind.online');
    }

    // Build both services up-front so a failure on the online side doesn't
    // leave a half-attached router.
    final offlineService = _createService(offline);
    final HybridServiceHandle onlineService;
    try {
      onlineService = _createService(online);
    } catch (_) {
      _bridge.destroyService(offlineService);
      rethrow;
    }

    // Detach + destroy any previously attached services before swapping in the
    // new pair (clear router slots first — see header UAF note), and retire the
    // previous policy's custom-filter predicates.
    _clearAndDestroyServices();
    _unregisterCustomFilters();

    final rcOff = _bridge.setOfflineService(
        _handle, offlineService, _descriptorBytes(offline));
    if (rcOff != RAC_SUCCESS) {
      _bridge
        ..destroyService(offlineService)
        ..destroyService(onlineService);
      throw StateError('set_offline_service_proto failed (rc=$rcOff)');
    }
    final rcOn = _bridge.setOnlineService(
        _handle, onlineService, _descriptorBytes(online));
    if (rcOn != RAC_SUCCESS) {
      _bridge
        ..setOfflineService(_handle, null, Uint8List(0))
        ..destroyService(offlineService)
        ..destroyService(onlineService);
      throw StateError('set_online_service_proto failed (rc=$rcOn)');
    }

    // Register custom-filter predicates with commons BEFORE installing the
    // policy bytes so the router can resolve each HybridFilter.custom name the
    // first time it filters. The router owns the eval — Dart only supplies the
    // named predicate.
    final customs = policy.customFilters;
    for (final filter in customs) {
      _bridge.registerCustomFilter(filter.name, filter.check);
    }
    final customNames = customs.map((f) => f.name).toList(growable: false);

    final rcPolicy = _bridge.setPolicy(_handle, policy.toProtoBytes());
    if (rcPolicy != RAC_SUCCESS) {
      for (final name in customNames) {
        _bridge.unregisterCustomFilter(name);
      }
      _bridge
        ..setOfflineService(_handle, null, Uint8List(0))
        ..setOnlineService(_handle, null, Uint8List(0))
        ..destroyService(offlineService)
        ..destroyService(onlineService);
      throw StateError('set_policy_proto failed (rc=$rcPolicy)');
    }

    _offline = offlineService;
    _online = onlineService;
    _customFilterNames = customNames;
  }

  /// Run one transcribe request through the router. The router applies the
  /// installed policy (filters → rank → invoke → fallback) in commons and
  /// returns the chosen backend's result plus the routing decision.
  ///
  /// [audioBytes] is file-encoded audio (wav/mp3/flac/...) OR raw PCM; each
  /// engine decodes per its own expectations. Throws [HybridTranscribeException]
  /// when the backend reports a non-zero rc, [StateError] on native failure.
  ///
  /// Returns the generated `HybridSttTranscribeResponse` (rc is always 0 on a
  /// successful return; the transcript is in `.text`, the routing decision in
  /// `.routing`). On the `.routing` metadata, `confidence` /
  /// `primaryConfidence` are absent (`hasConfidence()` / `hasPrimaryConfidence()`
  /// are false) when the engine surfaced no quality signal — guard on those
  /// accessors rather than reading the 0.0 default.
  pb.HybridSttTranscribeResponse transcribe(
    Uint8List audioBytes, {
    pb.HybridSttTranscribeOptions? options,
  }) {
    _ensureOpen();
    if (_offline == null || _online == null) {
      throw StateError('setPair() must be called before transcribe()');
    }

    // Pure pass-through: commons owns the entire routing decision AND the
    // raw-PCM16 → WAV payload normalisation (rac_stt_hybrid_router_proto.cpp).
    // Dart marshals the request and decodes the response; it does NOT
    // pre-filter candidates or toggle slots. HybridRoutingContext is empty on
    // the wire (device-state lives behind the vtable); it is still emitted
    // for a stable wire shape.
    final request = pb.HybridSttTranscribeRequest(
      audioBytes: audioBytes,
      context: pb.HybridRoutingContext(),
      options: options ?? pb.HybridSttTranscribeOptions(),
    );

    final responseBytes = _bridge.transcribe(_handle, request.writeToBuffer());
    return _decodeResponse(responseBytes);
  }

  /// Cancel an in-flight transcribe, if any. Best-effort: no STT engine
  /// exposes a cancel op today, so commons treats this as a no-op until one
  /// does. Mirrors Swift `HybridSTTRouter.cancel()` (HybridSTTRouter.swift:348).
  void cancel() {
    if (_closed) return;
    _bridge.cancelRouter(_handle);
  }

  /// Detach + destroy both services, unregister custom filters, and destroy the
  /// router handle. Idempotent.
  void close() {
    if (_closed) {
      return;
    }
    _closed = true;
    _unregisterCustomFilters();
    _clearAndDestroyServices();
    _bridge.destroyRouter(_handle);
    _handle = RacHandle.fromAddress(0);
  }

  // --- Internals -----------------------------------------------------------

  HybridServiceHandle _createService(HybridModel model) {
    final String? configJson;
    if (model.backend == HybridBackend.cloud) {
      final entry = CloudBackend.instance.lookup(model.id);
      if (entry == null) {
        throw StateError(
          "Cloud model id '${model.id}' not registered. Call "
          'RunAnywhere.hybrid.cloud.register(id, model, apiKey) at app startup.',
        );
      }
      configJson = cloudSttConfigJson(entry);
      // Cloud engine takes everything via config_json; no model path.
      return _bridge.createService(
        engineHint: model.backend.engineHint,
        modelIdOrPath: '',
        configJson: configJson,
      );
    }
    // Offline (sherpa): the model-registry path-resolution lives in commons;
    // the create routes through the same plugin registry (hint "sherpa").
    return _bridge.createService(
      engineHint: model.backend.engineHint,
      modelIdOrPath: model.id,
    );
  }

  Uint8List _descriptorBytes(HybridModel model) {
    final descriptor = pb.HybridModelDescriptor(
      modelId: model.id,
      modelType: model.kind.proto,
      backend: model.backend.proto,
      provider: model.provider,
    );
    return descriptor.writeToBuffer();
  }

  void _clearAndDestroyServices() {
    // Slot-clearing must precede service destruction (router holds raw
    // pointers — see rac_stt_hybrid_router.h UAF note).
    if (_handle != RacHandle.fromAddress(0)) {
      _bridge
        ..setOfflineService(_handle, null, Uint8List(0))
        ..setOnlineService(_handle, null, Uint8List(0));
    }
    _bridge
      ..destroyService(_offline)
      ..destroyService(_online);
    _offline = null;
    _online = null;
  }

  void _unregisterCustomFilters() {
    for (final name in _customFilterNames) {
      _bridge.unregisterCustomFilter(name);
    }
    _customFilterNames = const <String>[];
  }

  void _ensureOpen() {
    if (_closed) {
      throw StateError('HybridSttRouter is closed');
    }
  }

  pb.HybridSttTranscribeResponse _decodeResponse(Uint8List bytes) {
    final response = pb.HybridSttTranscribeResponse.fromBuffer(bytes);
    if (response.rc != 0) {
      throw HybridTranscribeException(
        response.rc,
        response.errorMsg.isEmpty
            ? 'Hybrid STT transcribe failed (rc=${response.rc})'
            : response.errorMsg,
      );
    }
    return response;
  }
}

/// Thrown by [HybridSttRouter.transcribe] when the chosen backend reports a
/// non-zero native rc. Carries the rc + the human-readable error message from
/// the HybridSttTranscribeResponse.
class HybridTranscribeException implements Exception {
  /// Build a transcribe exception.
  const HybridTranscribeException(this.code, this.message);

  /// Native `rac_result_t` from the failing backend.
  final int code;

  /// Human-readable failure reason.
  final String message;

  @override
  String toString() => 'HybridTranscribeException(code=$code): $message';
}
